### ISSUE TITLE
Create ES3 specific lib.d.ts

### ISSUE THREAD

**Issue by SergioMorchon** [microsoft/TypeScript#2410]
By targeting to ES3, the following code (within a class definition):

``` typescript
        setOptionDisable(option: HTMLOptionElement, operation: EnumOperations) {
            ko.applyBindingsToNode(option,
                {
                    disable: BI_OP.has(operation)? this.selectedNumbers().length >= 2 : true
                }, item);
        }
```

compiles, but the `item` variable is not defined. Does really `item` exists in a ES3 standard browser? I wasn't able to find it, so I treated this as a bug.

**Comment by danquirk** [comment]
You'll need to post more repro code. Simply using an undefined name like that in a small snippet is clearly an error. If there's actually a bug in the compiler or your code it's more involved than just that function definition.

**Comment by SergioMorchon** [comment]
On the lib.d.ts file is defined the following function:
``` typescript
declare function item(index: any): any;
```
But as seeing @ http://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%203rd%20edition,%20December%201999.pdf 15.1.1 Value Properties of the Global Object (page 77) and W3C http://www.w3schools.com/jsref/jsref_obj_global.asp , there isn't any ``` item ``` global function.

This misstake causes that if I use a non-defined ``` ítem ``` variable on my code, which of course doesn't exists, it will compile ok, but it shouldn't!

For example:
``` typescript
item();// must throw a compilation error. Instead, it is getting a runtime one on the browser because  it is not defined.
```

Same with ``` msClearImmediate ```, ``` btoa ```, ``` atob ```, etc.

I think I can explain it better... :S

**Comment by danquirk** [comment]
Ah, we could consider an ES3 specific lib.d.ts. In the meantime you should just remove those members from the copy of lib.d.ts you're building against or polyfill the implementation if you're targeting both ES3 and ES5 and want that functionality everywhere. We already see most people targeting ES5+ and that's only going to increase so this isn't likely to be super high priority to fix right now.

**Comment by RyanCavanaugh** [comment]
Would probably accept a PR on a good es3.lib.d.ts

**Comment by SergioMorchon** [comment]
This is very interesting at Enterprise environments that works with legacy browsers.

**Comment by mhegazy** [comment]
related to https://github.com/Microsoft/TypeScript/issues/6974

**Comment by ericdrobinson** [comment]
I understand that this is a relatively old Issue, but I would like to toss in a use-case for providing an es3-specific declaration lib. I find myself working with the [admittedly aging] [ExtendScript](https://en.wikipedia.org/wiki/ExtendScript) language, one that is primarily used for creating extensions to Adobe products. The language itself is [ECMAScript 3-based](http://stackoverflow.com/a/22537459/5285364).

It appears to be possible to write TypeScript such that compiled JavaScript can work with this language:

- Specify the `"es5"` lib in the compiler options (or whatever you wish to _write_ in).
  - Omit the `"dom"` lib.
- Set the `target` compiler option to `"es3"`. 
- Provide declaration files for other ExtendScript types.

However, it is not possible to write type-annotated ExtendScript in a TypeScript enabled environment. With a _lib.es3.d.ts_ declaration file option, this would be possible with very little change to the above configuration:

- Specify the `"es3"` lib in the compiler options. **[<-- Missing!]**
- Provide declaration files for other ExtendScript types.

As it stands, we can get _most_ of the way by specifying `"es5"`. The obvious pitfall with this strategy is that some features/APIs suggested with type completion will be erroneous/not supported by the target language.

Would be nice to have an `"es3"` lib option...

**Comment by mhegazy** [comment]
> Would be nice to have an "es3" lib option...

it sure would be. and this issue is tracking it. we would be happy to accept a PR for lib.es3.d.ts

**Comment by ericdrobinson** [comment]
Speaking for myself, I have no clue as to how the existing lib declaration files were built. Suggestions on how to quickly create a _lib.es3.d.ts_ file (perhaps by modifying the existing es5 version?) would be instructive. Furthermore, what would you suggest to use for the target spec? [This](https://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262,%203rd%20edition,%20December%201999.pdf)? If so, what portions are most important? I would _guess_ that **Chapter 15 - Native ECMAScript Objects** would be the primary focus. Are there any other important sections? Anything important to capture in **Chapter 8 - Types** or **Chapter 9 - Type Conversion**? Or are those things handled intrinsically by TypeScript already?

**Comment by mhegazy** [comment]
I would start by lib.es5.d.ts, and go through he spec remove/update declarations as necessary.

**Comment by ericdrobinson** [comment]
Out of curiosity, does _lib.es5.d.ts_ correspond to version [5.0](https://www.ecma-international.org/publications/files/ECMA-ST-ARCH/ECMA-262%205th%20edition%20December%202009.pdf) or [5.1](http://www.ecma-international.org/ecma-262/5.1/ECMA-262.pdf)?

**Comment by mhegazy** [comment]
5.1

**Comment by ericdrobinson** [comment]
@mhegazy Is there a resource that describes how the lib files are structured and why they are structured in the manner that they are? For example, why are all built-in types declared as `interface`s with equivalently named global `const`ants rather than `class`es?

This seems to make some kind of sense for `static`-like classes (e.g. `Math`). However, what about the basic `Object` class? Why split it up into an `Object` interface and an `ObjectConstructor` interface? It looks like this is . There doesn't appear to be a good description for this at the top of the [lib file itself](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es5.d.ts), nor in the declaration file [documentation](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html). Some googling led me to "construct signatures", which is somewhat covered in the "Difference between the static and instance sides of classes" section of the TypeScript [Interface documentation](https://www.typescriptlang.org/docs/handbook/interfaces.html). What's somewhat confusing is that TypeScript `class`es can have static methods/fields. So why not just use those?

One drawback(?) that I'm seeing with the interface method being employed is that [when using JavaScript, at least] the global namespace is filled not with classes as one might expect, but interfaces. I understand that JavaScript classes are not "real" classes, per se, but it is also somewhat surprising to see autocomplete turn up `ObjectConstructor` as a valid completion target.

Basically, why this:
```typescript
// Instance-side.
interface Blah {
    num: number;
}

// Static-side.
interface BlahConstructor {
    val: object;
    new (): Blah;
}

declare const Blah: BlahConstructor;
```
instead of this:
```typescript
declare class Blah {
    constructor();
    static val: object;
    num: number;
}
```
Convention? Or some deeper reason...?

(Please let me know if there is a better place to ask such questions. I ask them here to help anyone else considering taking on the task of creating the "es5" lib who may also not be terribly familiar with "why" behind the architecture.)

**Comment by mhegazy** [comment]
> @mhegazy Is there a resource that describes how the lib files are structured and why they are structured in the manner that they are? For example, why are all built-in types declared as interfaces with equivalently named global constants rather than classes?

classes are not callable. most of the built-in objects e.g. `Object`, `Array`, etc.. allow to be called as function as well as a constructor.

**Comment by mhegazy** [comment]
> One drawback(?) that I'm seeing with the interface method being employed is that [when using JavaScript, at least] the global namespace is filled not with classes as one might expect, but interfaces. I understand that JavaScript classes are not "real" classes, per se, but it is also somewhat surprising to see autocomplete turn up ObjectConstructor as a valid completion target.

that is unfortunate, and we tried to limit it to only required objects, e.g. DOM APIs do not behave the same way.

One thing to note, you need to keep the structure, for things like `ObjectConstructor`, the assumption when ppl use `--lib es3` is that they will get a smaller API surface area, not a completely different one.

**Comment by ericdrobinson** [comment]
# On Callable Core "classes"
>classes are not callable. most of the built-in objects e.g. Object, Array, etc.. allow to be called as function as well as a constructor.

Ahh, so that is what the last two lines of this are then:
```typescript
interface ObjectConstructor {
    new (value?: any): Object;
    (): any;
    (value: any): any;
    //...
}
```
I'd actually not previously understood why those were there. Makes much more sense.

## Small Bug in One of the Declarations?
I cross checked those with [the spec](https://www.ecma-international.org/ecma-262/5.1/index.html#sec-15.2.1.1). It seems to me that the one called without an argument should return an `Object`, no? E.g.:
```typescript
interface ObjectConstructor {
    //...
    (): Object;       // "return a new Object object"
    //...
}
```

# Un-implementable Interfaces?
Also, is the idea with the "callable interfaces" that they will simply make the interface un-implementable? I attempted to implement `ObjectConstructor` in a test class and got the following error:
```
'Class 'test' incorrectly implements interface 'ObjectConstructor'.
  Type 'test' provides no match for the signature '(): any''
```
I take it that this is "by design"?

# Maintaining the Current Structure with an `es3` Variant
>One thing to note, you need to keep the structure, for things like `ObjectConstructor`, the assumption when ppl use `--lib es3` is that they will get a smaller API surface area, not a completely different one.

Agreed. This was part of the impetus for requesting more information as to the structure of the declarations. Understanding why the structure is as it is will help anyone attempting to create a _lib.es3.d.ts_ file.

**Comment by ericdrobinson** [comment]
>Also, is the idea with the "callable interfaces" that they will simply make the interface un-implementable?
>...
>I take it that this is "by design"?

I'll answer my own question here. Yes: it appears that this is [by design](http://stackoverflow.com/a/16508581/5285364).

For anyone interested in more, [this blog entry](https://blogs.msdn.microsoft.com/typescript/2013/01/24/walkthrough-interfaces/) is extremely helpful. The [main documentation](https://www.typescriptlang.org/docs/handbook/interfaces.html) is also helpful if you look for "call signature" (under _Function Types_) and "construct signature" (under _Difference between the static and instance sides of classes_).

**Comment by mhegazy** [comment]
> I cross checked those with the spec. It seems to me that the two lines listed above may not be quite as accurate as they could be. I would E.g.:

sure. it just that ppl do not use `Object(1)` that often.

**Comment by mhegazy** [comment]
> I take it that this is "by design"?

yes. these are not meant to be implemented, but rather to describe JS objects that do exist.

**Comment by ericdrobinson** [comment]
>sure. it just that ppl do not use Object(1) that often.
I bet. Doesn't mean that it isn't technically wrong ;)

**Comment by mhegazy** [comment]
> @mhegazy I bet. Doesn't mean that it isn't technically wrong ;)

it is not wrong. it is just not that common.

**Comment by saschanaz** [comment]
Reopened my PR as #28372, if anyone still is interested.

**Comment by RyanCavanaugh** [comment]
Comments in the linked PR
